### DISCUSSION TITLE
Allow convenient object.Function() syntax for function parameters

### DISCUSSION THREAD

**Discussion by ymassad** [dotnet/csharplang#1917] [Language Ideas]
It would be really nice to be able to invoke Func and Action parameters using the object.Function() syntax. Consider the following example:

```
public static void DoSomething(
    Customer customer,
    ext Func<Customer, string> GetDescription,
    ext Func<Order, string> GetDescription)
{
    var desc1 = customer.GetDescription();

    Order order = ...

    var desc2 = order.GetDescription();

    ...

}
```

'ext' stands for extension. I am not sure if this is the best keyword though.

The DoSomething method takes two Func parameters. The special thing about these parameters is the 'ext' keyword applied to them.

The body of the method uses GetDescription on the Customer object as if it were an instance method. The same thing is true for Order.

In the example, two parameters have the same name. C# would allow this because they have the 'ext' keyword.

Under the hood, the C# compiler could name these Customer_GetDescription and Order_GetDescription and have special attributes applied to the parameters (or the method?) that contain the "real" names.

I see two benefits:

1) We get to name the parameters GetDescription instead of GetCustomerDescription and GetOrderDescription.

2) We get better experience by typing 'customer.' and then have IntelliSense help us complete the method invocation. I see this as getting best of both the functional programming way of composition and the OOP way of thinking about the object (the customer and the order) when deciding what to do next.

**UPDATE:**

The value of this feature is more visible when you invoke multiple functions in a single expression like this:

`TranslateXYZToABC(ConvertAddressToXYZ(GetMessageAddress(message)))`

With the proposed feature, it will look like this:

`message.GetAddress().ConvertToXYZ().TranslateToABC()`

Which looks much better to me.

**UPDATE:** When adding this issue I used the following syntax first, then I decided to change it to the syntax presented before:

```
public static void DoSomething(
    Customer customer,
    Func<this Customer, string> GetDescription,
    Func<this Order, string> GetDescription)
```

I think that the 'ext' syntax is much better because it allows us to use any delegate type and also because it does not require modifying the syntax for generic arguments.

**Comment by YairHalberstadt** [comment]
In other words you're asking for extension delegate types.

Unlike extension methods, this doesn't really give a powerful new way to extend existing objects ala Linq.

Instead it's a relatively minor bit of syntactic sugar which will only really be relevant locally to where the delegate is in scope.

I could imagine that there would indeed be a very few cases where I would consider using it.

But we have to consider every bit of syntactic sugar comes at a cost. People who've never seen this new syntax before will be confused by it. People will be trying to work out where this method comes from for example, or what `this` in a `Func` means. As such, even in those cases I probably wouldn't use it.

If it is added, it will have to be supported till the end of time, and no more powerful features could be added if their syntax conflicts with this.
 
As such I see this as too little gain, for too much downside, to be worth doing.

**Comment by jnm2** [comment]
I would never want a delegate parameter to be made to look like an instance member. Instance members indicate that they are an intrinsic part of the type, including extension methods. Method parameters do not qualify.

**Comment by ymassad** [comment]
@YairHalberstadt ,

> extension delegate types.

I am not sure exactly what you mean here. But I want to use existing delegate types without defining new ones.

> Instead it's a relatively minor bit of syntactic sugar which will only really be relevant locally to where the delegate is in scope.

True.

> People who've never seen this new syntax before will be confused by it.

Right. But this is true for any new feature.

>  People will be trying to work out where this method comes from for example, or what this in a Func means.

This is not a problem Today because the IDE can very easily help you get to the definition of the method (the function parameter in this case).

Thinking about the 'this' syntax, I think that the other syntax I suggested (the 'ext' syntax) is better. It allows the feature to work on any delegate type and does not require the generic argument syntax to be modified.

> and no more powerful features could be added if their syntax conflicts with this.

I can't image another feature that would use the exact same syntax.

**Comment by ymassad** [comment]
@jnm2 ,

> including extension methods

Extension methods are hardly an intrinsic part of the object. For example, are Select, SelectMany, and Where intrinsic parts of `IEnumerable<T>`? I think that it is just a convenient syntax.

**Comment by jnm2** [comment]
@ymassad The point of extension methods is to be able to give a type you don't own new intrinsic capabilities like `.Select`. It works and it is consistent; any time you have an `IEnumerable<>`, you know you can do `.Select`. It's one of the *reasons* you work with `IEnumerable<>`.

Method parameters are not at all good candidates to extend the set of members on types within the method body.

**Comment by ymassad** [comment]
@jnm2 , Ok. Then maybe I used the wrong name of the feature. It should be something like: convenient object.Function() syntax for function parameters.

**Comment by jnm2** [comment]
I'm objecting to the `instance.Member` syntax, not the title of this issue.  The members available on a given type should not differ from point to point within the same class.

`methodParam(instance)` is appropriate, `instance |> methodParam` would be appropriate, but `instance.methodParam()` is not appropriate from my perspective.

**Comment by svick** [comment]
What if extension local functions (https://github.com/dotnet/csharplang/issues/850) were allowed? With that, you could write:

```c#
public static void DoSomething(
    Customer customer,
    Func<Customer, string> getCustomerDescription,
    Func<Order, string> getOrderDescription)
{
    string GetCustomerDescription(this Customer c) => getCustomerDescription(c);
    string GetOrderDescription(this Order o) => getOrderDescription(o);

    var desc1 = customer.GetCustomerDescription();

    Order order = ...

    var desc2 = order.GetOrderDescription();

    ...

}
```

It's quite a lot of extra syntax compared with this proposal, but it's also more widely useful and builds on existing C# syntax (`this` on a parameter) instead of introducing brand new syntax (`this` on a type argument of a generic type).

**Comment by ymassad** [comment]
@svick , In your example, you still need to name the local functions GetCustomerDescription and GetOrderDescription instead of simply GetDescription. Also, as you noted, it requires more work.

I agree with you regarding the 'this' syntax. I have suggested another syntax at the end of the original post.

**Comment by ymassad** [comment]
@jnm2 , you have a point. I think that my proposal makes the member.Method() syntax have a different meaning than what it used to mean. Extension methods have changed the meaning a little bit (consider that two invocations with the exact same type and member name in different files can mean different things based on the 'using' declarations used). This proposal moves this a bit further. IDEs can help very much by highlighting the function parameter invocation in a different way.

**Comment by theunrepentantgeek** [comment]
> IDEs can help very much by highlighting the function parameter invocation in a different way.

Not everyone uses an IDE. Particularly with .NET Core, there's a considerable rise in folks using sophisticated text editors.

IDE assistance can make language features easier to use - but language features shouldn't be dependent on IDE assistance to be understandable.

**Comment by ymassad** [comment]
@theunrepentantgeek , what you say can be said about almost any C# feature.

For example, if you see this:

`Customer.DoSomething();`

What is `Customer`? Is it a class name? a variable name? a field name? a parameter?

With an IDE, things are better. Without an IDE, you can get to what you need to know but it requires more knowledge of the code you are reading and the language itself.

**Comment by jnm2** [comment]
>  Is it a class name? a variable name? a field name? a parameter?

Maybe a property name, but no way would I use pascal casing for variables, fields or parameters.

**Comment by theunrepentantgeek** [comment]
> What is `Customer`?

If the codebase adheres to the norms of the wider C# ecosystem, it _should_ be either a type name or a property. It's unlikely to be a variable, field or parameter.

That said, I suspect you're missing my point @ymassad.

Language features shouldn't _require_ IDE support to be understood; they need to stand on their own.

**Comment by ymassad** [comment]
@jnm2 , You can never know. Other developers work on the same project.

What about this:

`customer.DoSomething();`

In summary, when you work with a language, you need to know the features.

**Comment by ymassad** [comment]
@theunrepentantgeek , I am not saying that an IDE is required. I am only saying that it helps.

If a developer is working on a C# project and he/she sees an `ext` function parameter, they can expect that it will be used within the method (in the object.Function() way). If they don't know what 'ext' is, then can search on it and learn it.

Also, methods are expected to be relatively short, so the parameter declaration and its usage would almost certainly be seen at once (on the same screen, without scrolling).

**Comment by ymassad** [comment]
@theunrepentantgeek , think about the `yield` keyword. If a developer sees it, they need to learn it. What about lambdas? I remember the first time I saw one, I was confused. But I went to google and searched it.

**Comment by Austin-bryan** [comment]
@ymassad 
> For example, if you see this:
> Customer.DoSomething();
> What is Customer? Is it a class name? a variable name? a field name? a parameter

Thankfully we have conventions to help us out. It's likely a class, property or public field, not likely to be a local, or parameter. If I saw `customer.DoSomething();` I'd assume it to be a local or a parameter. 

Besides the point, not even in real life languages can you take things out of context and expect them to still make sense, not even individual words, as they can have multiple meanings. 

Also, if it's already tough to know what exactly is going on with `Customer.DoSomething();`, you're idea is to make that problem even *more* ambigious, by now not letting us know if it's an `ext` method parameter? Saying, "It's already bad, so why not make it worse" sounds like a logical fallacy.

C# language devs say that a feature must be overwelmingly postiitive to be considered, but this is just turning, `DoSomethingOrder(order);` to `order.DoSomething();`. Barely a difference, but can cause even more confusion and will barely even be used.

**Comment by ymassad** [comment]
@willard720 , I think you have made good points.

The way I write programs is that I have data objects that contain no behavior (they just model data), and then behavior 'objects' or functions that contain only behavior.

When doing such separation, and when implementing this feature, `customer.GetDescription()` would be easily understood to mean invoking the GetDescription function parameter and not an instance method (because there are none by design).

I guess this feature might make more sense in a more functional language.

**Comment by Austin-bryan** [comment]
@ymassad So `customer` doesn't have any methods of it's own? Well, seems to me, there's your problem. You're wanting to make a language feature because of you're rare choice in design. We are in an object orientated language, so you should use instance methods, it makes life easier. 

Another option, if you *really* need separation between the data and methods so that you can change the implementation of those methods at run time and you want method syntax, instead of function syntax, you can create something like this:

```csharp
public class BaseClass
{
    public string Description { get; private set; }
}
public class Customer : BaseClass { /* Data only */ }
public class Order : BaseClass { /* Data only */ }

public abstract class Behavior<BaseClass>
{
    public abstract string GetDescription(); 
}
public class CustomerBehavior : Behavior<Customer>
{
    public readonly Customer customer;
    
    public CustomerBehavior(Customer customer) => this.customer = customer;
    public override string GetDescription() => customer.Description;
}
public class OrderBehavior : Behavior<Order>
{
    public readonly Order order;
    
    public OrderBehavior(Order order) => this.order = order;
    public override string GetDescription() => order.Description;
}

public class NewCustomerBehavior : CustomerBehavior
{
    public NewCustomerBehavior(Customer customer) : base(customer) { }
    public override string GetDescription() => "No description available";
}
public class NewOrderBehavior : OrderBehavior
{
    public NewOrderBehavior(Order order) : base(order) { }
    public override string GetDescription() => "No description available";
}

...
    
public void Foo()
{
    var customer = new Customer();
    var order = new Order();
    
    DoSomething(new CustomerBehavior(customer), new OrderBehavior(order));
    DoSomething(new NewCustomerBehavior(customer), new NewOrderBehavior(order));
}
public void DoSomething(CustomerBehavior customerBehavior, OrderBehavior orderBehavior)
{
    customerBehavior.GetDescription();
    orderBehavior.GetDescription();
}
```

So basically, you have the `BaseClass` which `Customer` and `Order` inherit from, and all the other data-only classes. Then there is `Behavior<BaseClass>` which defines a `GetDescription` method. From there is a `CustomerBehavior` which inherits from `Behavior<Customer>` and an `OrderBehavior` which inherits from `Behavior<Order>`. 

Those two classes define the base operations for the customer and order behavior, but they themselves can have child classes, such as `NewCustomerBehavior` and `NewOrderBehavior` which do something else for `GetDescription`. This means that to get the proper method you want, you just need to create the correct class that you need, and you can even use a `BehaviorFactory` for that, which can have a `CustomerBehaviorFactory` child. 

Another cool thing about this design is that it encapsulates the methods for reuse later on. 

I'm pretty sure there is already a design pattern similar to this, but it's likely to be better. It's something to do with seperating the method implementation with the class itself. I'm also pretty sure it's possible to do that with having the method still attached to the class by doing something like having `Customer` have an internal `CustomerBehavior` that `Customer` forwards its methods too, and then it switches the particular instance of `CustomerBehavior` whenever it's `enum` `Status` is changed, for example, so you can fully abstract that process. 

Regardless, there's lots of options, and therefore makes it not necessary to have the `ext` delegate, as you can just use a design pattern.

**Comment by iam3yal** [comment]
@ymassad I'm really confused it looks like it's really forced so I'm really curious, what led you to write this proposal? I mean why would something like the following wouldn't work for you?
```
public static void DoSomething(
	Customer customer,
	DescriptionManager manager
{
	var desc1 = manager.GetDescription<Customer>();

	Order order = ...

	var desc2 = manager.GetDescription<Order>();

    ...
}
```

**Comment by DavidArno** [comment]
@willard720,
> So `customer` doesn't have any methods of it's own? Well, seems to me, there's your problem. You're wanting to make a language feature because of you're rare choice in design. We are in an object orientated language, so you should use instance methods, it makes life easier.

[As @CyrusNajmabadi recently put it](https://github.com/dotnet/csharplang/issues/84#issuecomment-427526080), "*C# is not OO. C# is a multi-paradigm language. It embraces many different styles, including: imperative, OO, functional, dynamic, and many others. It is vehemently non-ideological.*"

Like @ymassad, most of my C# code these  days consists of pure (ie side effect-free) static methods and data-only objects that contain no methods of their own. Sure, it's still a minority way of writing C# code, but it's far from rare these days. And to my mind, replacing the simple elegance of a couple of delegates with heavy use of inheritance does indeed have a name: it's called an *anti-pattern*.


@ymassad,

The problem I have with your proposal is summed up by the claim,
> 1. We get to name the parameters GetDescription instead of GetCustomerDescription and GetOrderDescription.

I don't see this as an advantage. I'd far rather see a simple delegate parameter and `GetCustomerDescription(customer)` than a local extension method that let's me express it as `customer.GetDescription()`. I just don't see the benefit of extension methods in this scenario. As an aside though, the "data then function" order does have merit, but I think this would be better solved with [the forward pipe operator](https://github.com/dotnet/csharplang/issues/96):

```cs
var desc1 = customer |> GetCustomerDescription();
```

**Comment by ymassad** [comment]
@eyalsk , DescriptionManager in this case is too heavy a parameter to pass. It gives the method too much access. If I look at the `DoSomething` method signature I will have less knowledge of what the DoSomething method does (because it could use DescriptionManager to get the description of anything, not just Customer and Order). Also, how can I customize DescriptionManager? Passing a function (as a lambda or a method group) is much more customizable.

**Comment by ymassad** [comment]
@DavidArno , the suggestion you made I think gives us half the benefit. I assume that once you type `customer |>`, IntelliSense will suggest you pass it to `GetCustomerDescription()`.

The other half is related to shortening the function parameter name.

Another proposal would be to allow parameters whose types are delegates to have the same name (without the ext keyword), of course as long as the delegate **parameter** types (the parameters defined by the delegate) are different (e.g. like the ones in my example). This will give us half the benefit. And When the forward pipe operator is implemented, that would give us the other half.

**Comment by ymassad** [comment]
@willard720 , I think that the solution you provided is very complex compared to the one I use.

**Comment by HaloFour** [comment]
It's an interesting proposal but I can't see it being of much benefit to anyone except those that follow this specific design pattern, which I can't imagine is remotely common.  Pretty much all of the reasons for wanting the language change is due to adherence to a very strict and self-impose set of restrictions, which preclude a swath of existing language features.  I also find it pretty weird that you want to follow functional idioms but that you want to do so with OOP syntax.

**Comment by iam3yal** [comment]
@ymassad 

> DescriptionManager in this case is too heavy a parameter to pass. It gives the method too much access. If I look at the DoSomething method signature I will have less knowledge of what the DoSomething method does (because it could use DescriptionManager to get the description of anything, not just Customer and Order).

This is only because you didn't provide enough information in your post, in fact, you don't even mention the alternatives and the issues associated with them but to address the problem you can add an extension method named `GetDescription` to `Customer` and `Order` and then make them call the `DescriptionManager` and you will have almost what you want/need.

> Also, how can I customize DescriptionManager

`DescriptionManager.AddDescriptor<TKey>(string desc)`  or `DescriptionManager.AddDescriptor<TKey>(Func<TKey, string> descFunc)`

**Comment by iam3yal** [comment]
@ymassad I know that my suggestion is far from perfect but at least now I can see where this proposal might be useful.

**Comment by ymassad** [comment]
@HaloFour ,

>  I also find it pretty weird that you want to follow functional idioms but that you want to do so with OOP syntax.

You are right. I like how functions compose better than objects. But I see the object.Function() syntax very convenient in many cases.

**Comment by ymassad** [comment]
@eyalsk ,

>  but to address the problem you can add an extension method named GetDescription to Customer and Order and then make them call the DescriptionManager

This is not the same as passing the delegates to the method. In my post, the two delegate parameters can have different values on each call to DoSomething.

**Comment by Austin-bryan** [comment]
@ymassad There's really hardly a difference between `foo.Method()` and `function(foo)`, except for semantics. `foo.Method()` implies `Method()` is a part of `foo`, which makes it completely counter intuitive for it to be a delegate. It's therefore hiding that information.

And while my solution involves more typing, it also has more benefits such as code reuse. Besides that, the fact that it's possible to get that result with other design patterns currently makes this feature much less important, espically when the problem isn't common at all. `foo { get { return bar; } }` was so common that we had to have `foo => bar`, despite not allowing for anything new to be done. This isn't the case with this feature as I can't imagine ever needing this feature.

**Comment by ymassad** [comment]
@willard720 ,

There would be a difference if you have multiple functions, e.g.:

`message.GetAddress().ConvertToXYZ().TranslateToABC()`

looks much better to me than:

`TranslateXYZToABC(ConvertAddressToXYZ(GetMessageAddress(message)))`

**Comment by Austin-bryan** [comment]
@ymassad There's already a feature proposed that would fix that and is widely approved of. I'm pretty sure it's already a champion, I just can't find it. The syntax will look something like this:

```csharp
TranslateToABC() |> ConvertToXYZ() |> GetAddress() |> message;
```

and does the same thing. This is already a feature in other languages so it has a proper term. @HaloFour  Do you which feature I'm talking about? I'm not sure if I got the order of the method calls right or not. 

Either way, it's very elegant and solves that problem.

**Comment by yaakov-h** [comment]
#96?

**Comment by DavidArno** [comment]
@willard720,

You appear to have the expression back to front. The forward pipe operator (see #74 and #96) would transform:
```cs
TranslateXYZToABC(ConvertAddressToXYZ(GetMessageAddress(message)));
```
into:
```cs
message |> GetMessageAddress() |> ConvertAddressToXYZ() |> TranslateXYZToABC();
```

**Comment by DavidArno** [comment]
With regard to the above, if I may indulge in a bit of cheeky self-promotion, a library I maintain, [SuccincT](https://github.com/DavidArno/SuccincT/wiki/PipeOperators), provides basic forward pipe operations via the `Into()` extension method, allowing it to be written as:
```cs
message.Into(GetMessageAddress).Into(ConvertAddressToXYZ).Into(TranslateXYZToABC);
```
It's not as elegant as having the operator, and obviously has the performance overheads associated with delegates, but some might like it 😀

**Comment by Austin-bryan** [comment]
@DavidArno Pipe operators, that's what I was thinking of.

And the extension methods syntax isn't bad either! Though isn't there gonna be a performance hit from sending the delegates?

**Comment by DavidArno** [comment]
@willard720, good point re performance. I updated my comment to admit that.

**Comment by jnm2** [comment]
@willard720

> Though isn't there gonna be a performance hit from sending the delegates?

You know there is. No worse than LINQ though, so it hardly seems worth worrying about. The rule about optimizing measured hot paths still applies.

**Comment by Austin-bryan** [comment]
@jnm2 I assumed it wouldn't be a massive performance. I made a similar thing once and it ended up being around 10x slower, so it was around 0.0001s, instead of 0.00001s to use it like a million times in a loop. So basically it isn't a big enough difference to worth mentioning and I assume a similar difference will be present here.
